File: app/src/main/java/de/thu/recipebook/activities/RecipeDetailsActivity.java
package de.thu.recipebook.activities;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.ShareActionProvider;
import androidx.core.view.MenuItemCompat;

import android.content.ContentValues;
import android.content.Intent;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.provider.BaseColumns;
import android.speech.tts.TextToSpeech;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.Locale;

import de.thu.recipebook.databases.FavoritesDbHelper;
import de.thu.recipebook.runnables.DeleteRecipeRunnable;
import de.thu.recipebook.runnables.FetchRecipeDetailsRunnable;
import de.thu.recipebook.R;
import de.thu.recipebook.models.Recipe;

public class RecipeDetailsActivity extends AppCompatActivity {
    public String ADD_TO_FAVORITES = "Add To Favorites";
    public String REMOVE_FROM_FAVORITES = "Remove From Favorites";

    //    Ex. 1
//    private RecipeDatabase recipeDatabase;
    private ShareActionProvider shareActionProvider;
    private FetchRecipeDetailsRunnable fetchRecipeDetailsRunnable;
    private DeleteRecipeRunnable deleteRecipeRunnable;
    private FavoritesDbHelper favoritesDbHelper;
    private TextToSpeech textToSpeech;

    private MenuItem shareItem;
    private MenuItem editItem;
    private MenuItem deleteItem;
    private MenuItem favoriteItem;

    private Recipe recipe;

    private TextView nameTextView;
    private TextView countryTextView;
    private TextView countryLabelTextView;
    private TextView ingredientsTextView;
    private TextView ingredientsLabelTextView;
    private TextView instructionsTextView;
    private TextView instructionsLabelTextView;
    private Button listenButton;
    private ImageView imageView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_recipe_details);

//        Exercise 5
//        StrictMode.ThreadPolicy policy = new StrictMode.ThreadPolicy.Builder().permitAll().build();
//        StrictMode.setThreadPolicy(policy);
        deleteRecipeRunnable = new DeleteRecipeRunnable(this);
        fetchRecipeDetailsRunnable = new FetchRecipeDetailsRunnable(this);
        new Thread(fetchRecipeDetailsRunnable).start();

        favoritesDbHelper = new FavoritesDbHelper(this);
        textToSpeech = new TextToSpeech(getApplicationContext(), status -> {
            if (status != TextToSpeech.ERROR) {
                textToSpeech.setLanguage(Locale.UK);
            }
        });
//        Ex. 1
//        recipeDatabase = new RecipeDatabase();
//        currentRecipe = recipeDatabase.getNextRecipe(currentRecipe);

        nameTextView = findViewById(R.id.text_view_name);
        countryTextView = findViewById(R.id.text_view_country);
        countryLabelTextView = findViewById(R.id.text_view_country_label);
        ingredientsTextView = findViewById(R.id.text_view_ingredients);
        ingredientsLabelTextView = findViewById(R.id.text_view_ingredients_label);
        instructionsTextView = findViewById(R.id.text_view_instructions);
        instructionsLabelTextView = findViewById(R.id.text_view_instructions_label);
        listenButton = findViewById(R.id.button_listen);
        imageView = findViewById(R.id.image_view_picture);
    }

//    Ex. 1
//    public void nextButtonOnClick(View view) {
//        currentRecipe = recipeRepository.getNextRecipe(currentRecipe);
//        setRecipe();
//    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.recipe_details_menu, menu);
        shareItem = menu.findItem(R.id.action_share);
        editItem = menu.findItem(R.id.edit_entry);
        deleteItem = menu.findItem(R.id.delete_entry);
        favoriteItem = menu.findItem(R.id.favorite_entry);
        shareActionProvider = (ShareActionProvider)
                MenuItemCompat.getActionProvider(shareItem);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.favorite_entry:
                SQLiteDatabase db = favoritesDbHelper.getWritableDatabase();
                if (item.getTitle().equals(ADD_TO_FAVORITES)) {
                    ContentValues values = new ContentValues();
                    values.put(BaseColumns._ID, recipe.getId());
                    db.insert(FavoritesDbHelper.FAVORITES_TABLE, null, values);
                    item.setTitle(REMOVE_FROM_FAVORITES);
                } else {
                    db.delete(FavoritesDbHelper.FAVORITES_TABLE, BaseColumns._ID + "=?", new String[]{recipe.getId()});
                    item.setTitle(ADD_TO_FAVORITES);
                }
                break;
            case R.id.edit_entry:
                Intent updateRecipeIntent = new Intent(this, SaveRecipeActivity.class);
                updateRecipeIntent.putExtra("id", recipe.getId());
                updateRecipeIntent.putExtra("name", recipe.getName());
                updateRecipeIntent.putExtra("country", recipe.getCountry());
                updateRecipeIntent.putExtra("ingredients", recipe.getIngredients());
                updateRecipeIntent.putExtra("instructions", recipe.getInstructions());
                updateRecipeIntent.putExtra("image", recipe.getImage());

                startActivity(updateRecipeIntent);
                break;
            case R.id.delete_entry:
                new Thread(deleteRecipeRunnable).start();
        }
        return super.onOptionsItemSelected(item);
    }

    public void setRecipe(Recipe recipe) {
        nameTextView.setText(recipe.getName());
        countryTextView.setText(recipe.getCountry());
        countryLabelTextView.setVisibility(TextView.VISIBLE);
        ingredientsTextView.setText(recipe.getIngredients());
        ingredientsLabelTextView.setVisibility(TextView.VISIBLE);
        instructionsTextView.setText(recipe.getInstructions());
        instructionsLabelTextView.setVisibility(TextView.VISIBLE);
        listenButton.setVisibility(TextView.VISIBLE);

        if (recipe.getImage() != null) {
            Bitmap bitmap = BitmapFactory.decodeByteArray(recipe.getImage(), 0, recipe.getImage().length);
            imageView.setImageBitmap(bitmap);
        } else {
            imageView.setImageResource(R.drawable.dish_picture_placeholder);
        }

        this.recipe = recipe;
        if (recipe.isCreator()) {
            editItem.setVisible(true);
            deleteItem.setVisible(true);
        }

        setFavoriteMenuItemText();
        setShareText();
    }

    public void countryTextViewOnClick(View view) {
        String country = ((TextView) view).getText().toString();
        Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse("geo:0,0`?q=" + country));
        startActivity(intent);
    }

    public void listen(View view) {
        Button button = (Button) view;
        if (textToSpeech.isSpeaking()) {
            textToSpeech.stop();
            button.setText("Play");
        } else {
            textToSpeech.speak(recipe.toString(), TextToSpeech.QUEUE_FLUSH, null, null);
            button.setText("Stop");
        }
    }

    private void setShareText() {
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        shareActionProvider.setShareIntent(shareIntent);
        shareIntent.putExtra(Intent.EXTRA_TEXT, recipe.toString());
        shareItem.setVisible(true);
    }

    private void setFavoriteMenuItemText() {
        SQLiteDatabase db = favoritesDbHelper.getReadableDatabase();
        Cursor c = db.query(FavoritesDbHelper.FAVORITES_TABLE, new String[]{BaseColumns._ID},
                BaseColumns._ID + " = ?", new String[]{recipe.getId()}, null, null, null);
        favoriteItem.setVisible(true);
        if (c.getCount() > 0) {
            favoriteItem.setTitle(REMOVE_FROM_FAVORITES);
        }

    }
}

File: Exercise 8/Kotlin/src/main/java/de/thu/recipebook/activities/RecipeDetailsActivity.kt
package de.thu.recipebook.activities

import android.content.ContentValues
import android.content.Intent
import android.database.sqlite.SQLiteDatabase
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.provider.BaseColumns
import android.speech.tts.TextToSpeech
import android.speech.tts.TextToSpeech.OnInitListener
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.ShareActionProvider
import androidx.core.view.MenuItemCompat
import de.thu.recipebook.R
import de.thu.recipebook.databases.FavoritesDbHelper
import de.thu.recipebook.models.Recipe
import de.thu.recipebook.runnables.DeleteRecipeRunnable
import de.thu.recipebook.runnables.FetchRecipeDetailsRunnable
import java.util.*

class RecipeDetailsActivity : AppCompatActivity() {
    var ADD_TO_FAVORITES = "Add To Favorites"
    var REMOVE_FROM_FAVORITES = "Remove From Favorites"

    private var recipe: Recipe? = null
    private var shareActionProvider: ShareActionProvider? = null
    private var favoritesDbHelper: FavoritesDbHelper? = null
    private var textToSpeech: TextToSpeech? = null

    private var fetchRecipeDetailsRunnable: FetchRecipeDetailsRunnable? = null
    private var deleteRecipeRunnable: DeleteRecipeRunnable? = null

    private var shareItem: MenuItem? = null
    private var editItem: MenuItem? = null
    private var deleteItem: MenuItem? = null
    private var favoriteItem: MenuItem? = null

    private var nameTextView: TextView? = null
    private var countryTextView: TextView? = null
    private var ingredientsTextView: TextView? = null
    private var instructionsTextView: TextView? = null
    private var imageView: ImageView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_recipe_details)

        deleteRecipeRunnable = DeleteRecipeRunnable(this)
        fetchRecipeDetailsRunnable = FetchRecipeDetailsRunnable(this)
        Thread(fetchRecipeDetailsRunnable).start()

        favoritesDbHelper = FavoritesDbHelper(this)
        textToSpeech = TextToSpeech(applicationContext) { status: Int ->
            if (status != TextToSpeech.ERROR) {
                textToSpeech!!.language = Locale.UK
            }
        }

        nameTextView = findViewById(R.id.text_view_name)
        countryTextView = findViewById(R.id.text_view_country)
        ingredientsTextView = findViewById(R.id.text_view_ingredients)
        instructionsTextView = findViewById(R.id.text_view_instructions)
        imageView = findViewById(R.id.image_view_picture)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.recipe_details_menu, menu)
        shareItem = menu.findItem(R.id.action_share)
        editItem = menu.findItem(R.id.edit_entry)
        deleteItem = menu.findItem(R.id.delete_entry)
        favoriteItem = menu.findItem(R.id.favorite_entry)
        shareActionProvider = MenuItemCompat.getActionProvider(shareItem) as ShareActionProvider
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.edit_entry -> {
                val updateRecipeIntent = Intent(this, SaveRecipeActivity::class.java)
                updateRecipeIntent.putExtra("recipe", recipe)
                startActivity(updateRecipeIntent)
            }
            R.id.delete_entry -> Thread(deleteRecipeRunnable).start()
            R.id.favorite_entry -> {
                val db = favoritesDbHelper!!.writableDatabase
                if (item.title == ADD_TO_FAVORITES) {
                    val values = ContentValues()
                    values.put(BaseColumns._ID, recipe!!.id)
                    db.insert(FavoritesDbHelper.FAVORITES_TABLE, null, values)
                    item.title = REMOVE_FROM_FAVORITES
                } else {
                    db.delete(
                        FavoritesDbHelper.FAVORITES_TABLE, BaseColumns._ID + "=?", arrayOf(
                            recipe!!.id
                        )
                    )
                    item.title = ADD_TO_FAVORITES
                }
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onResume() {
        super.onResume()
        Thread(fetchRecipeDetailsRunnable).start()
    }

    fun setRecipe(recipe: Recipe) {
        nameTextView!!.text = recipe.name
        countryTextView!!.text = recipe.country
        ingredientsTextView!!.text = recipe.ingredients
        instructionsTextView!!.text = recipe.instructions

        if (recipe.image != null) {
            val bitmap = BitmapFactory.decodeByteArray(recipe.image, 0, recipe.image!!.size)
            imageView!!.setImageBitmap(bitmap)
        } else {
            imageView!!.setImageResource(R.drawable.dish_picture_placeholder)
        }

        this.recipe = recipe

        if (recipe.isCreator!!) {
            editItem!!.isVisible = true
            deleteItem!!.isVisible = true
        }

        setShareText()
        setFavoriteMenuItemText()
    }

    fun countryTextViewOnClick(view: View) {
        val country = (view as TextView).text.toString()
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse("geo:0,0`?q=$country"))
        startActivity(intent)
    }

    fun listen(view: View) {
        val button = view as Button
        if (textToSpeech!!.isSpeaking) {
            textToSpeech!!.stop()
            button.text = "Play"
        } else {
            textToSpeech!!.speak(recipe.toString(), TextToSpeech.QUEUE_FLUSH, null, null)
            button.text = "Stop"
        }
    }

    private fun setShareText() {
        val shareIntent = Intent(Intent.ACTION_SEND)
        shareIntent.type = "text/plain"
        shareActionProvider!!.setShareIntent(shareIntent)
        shareIntent.putExtra(Intent.EXTRA_TEXT, recipe.toString())
    }

    private fun setFavoriteMenuItemText() {
        val db = favoritesDbHelper!!.readableDatabase
        val c = db.query(
            FavoritesDbHelper.FAVORITES_TABLE, arrayOf(BaseColumns._ID),
            BaseColumns._ID + " = ?", arrayOf(recipe!!.id), null, null, null
        )
        favoriteItem!!.isVisible = true
        if (c.count > 0) {
            favoriteItem!!.title = REMOVE_FROM_FAVORITES
        }
    }
}